SU2_CFD/include/gradients/computeGradientsGreenGauss.hpp

```diff
@@ -0,0 +1,189 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "../../../Common/include/omp_structure.hpp"
+
+
+/*!
+ * \brief Compute the gradient of a field using the Green-Gauss theorem.
+ * \note Gradients can be computed only for a contiguous range of variables, defined
+ *       by [varBegin, varEnd[ (e.g. 0,1 computes the gradient of the 1st variable).
+ *       This can be used, for example, to compute only velocity gradients.
+ * \note The function uses an optional solver object to perform communications, if
+ *       none (nullptr) is provided the function does not fail (the objective of
+ *       this is to improve test-ability).
+ * \param[in] solver - Optional, solver associated with the field (used only for MPI).
+ * \param[in] kindMpiComm - Type of MPI communication required.
+ * \param[in] kindPeriodicComm - Type of periodic communication required.
+ * \param[in] geometry - Geometric grid properties.
+ * \param[in] config - Configuration of the problem, used to identify types of boundaries.
+ * \param[in] field - Generic object implementing operator (iPoint, iVar).
+ * \param[in] varBegin - Index of first variable for which to compute the gradient.
+ * \param[in] varEnd - Index of last variable for which to compute the gradient.
+ * \param[out] gradient - Generic object implementing operator (iPoint, iVar, iDim).
+ */
+template<class FieldType, class GradientType>
+void computeGradientsGreenGauss(CSolver* solver,
+                                MPI_QUANTITIES kindMpiComm,
+                                PERIODIC_QUANTITIES kindPeriodicComm,
+                                CGeometry& geometry,
+                                CConfig& config,
+                                const FieldType& field,
+                                size_t varBegin,
+                                size_t varEnd,
+                                GradientType& gradient)
+{
+  size_t nPointDomain = geometry.GetnPointDomain();
+  size_t nDim = geometry.GetnDim();
+
+#ifdef HAVE_OMP
+  constexpr size_t OMP_MAX_CHUNK = 512;
+
+  size_t chunkSize = computeStaticChunkSize(nPointDomain,
+                     omp_get_max_threads(), OMP_MAX_CHUNK);
+#endif
+
+  /*--- Start OpenMP parallel section. ---*/
+
+  SU2_OMP_PARALLEL
+  {
+    /*--- For each (non-halo) volume integrate over its faces (edges). ---*/
+
+    SU2_OMP_FOR_DYN(chunkSize)
+    for (size_t iPoint = 0; iPoint < nPointDomain; ++iPoint)
```

Comment:
This divides the loop over however many threads arrive here, without the worksharing construct the entire loop would be performed by each thread.
For reasons of spatial locality it is beneficial to divide the loop in chunks instead of individual iterations, in this case those chunks are assigned to each thread dynamically, this helps with load balancing but has some overhead. That is why you also see static work division for light loops (e.g. setting an array to 0).

Reply:
what is the `chunkSize` dependent on? Is it hardware? number of variables?

Does simply adding `SU2_OMP_FOR_DYN(chunkSize)` before the for loop automatically run the loop in chunks? (assuming you add relevant includes and calls beforehand)

Does anything new need to be done in how the code is compiled or run to ensure that multiple threads are used?

Reply:
The size is usually based on amount of work per loop iteration, if the loop is heavy a small size can be used to improve load balancing, otherwise a relatively large size needs to be used to amortize the parallelization overheads. Currently most sizes are determined such that each thread gets on average the same number of chunks but each chunk is no larger than a maximum size, there is an utility function in omp_structure.hpp for this.

Yes, distributing the work over the threads that arrive at that location, if only one arrives the code still executes properly.

The build system was updated in #843 since that introduces the first solver that is completely hybrid parallel, there is no advantage in using the feature for fluid solvers yet.
Basically you add -Dwith-omp=true to meson, the number of threads per rank can be specified with the -t option when calling SU2_CFD, e.g. `mpirun -n 2 --bind-to numa SU2_CFD -t 8 config.cfg` if -t is omitted the number of threads is whatever is defined for the environment.